### PR TITLE
Shading work

### PR DESCRIPTION
**Pull request by est77**
- Added Glass BSDF.
- Removed some OSL closures now that the replacements are in place.
- Added test scenes for the new BSDFs, updated OSL test scenes.

### CODE REVIEW

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 1]
```diff
@@ -1289,9 +1114,57 @@ BOOST_STATIC_ASSERT(sizeof(CompositeSurfaceClosure) <= InputEvaluator::DataSize)
...
+
+        m_ior_cdf[m_num_iors - 1] = 1.0;
+    }
+}
+
+void CompositeSurfaceClosure::add_ior(
+    const foundation::Color3f&  weight,
+    const double                ior)
+{
+    // We use the luminance of the weight as the ior weight.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: *IOR*

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 2]
```diff
@@ -1289,9 +1114,57 @@ BOOST_STATIC_ASSERT(sizeof(CompositeSurfaceClosure) <= InputEvaluator::DataSize)
...
+  : CompositeClosure()
+  , m_num_iors(0)
 {
     process_closure_tree(ci, original_shading_basis, Color3f(1.0f));
     compute_cdf();
+
+    // Build the IOR CDF in place if needed.
+    if (m_num_iors == 0)
+    {
+        m_num_iors = 1;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Where is `m_num_iors` increased? Can it really be larger than 1?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+                m_mdf.reset(new GGXMDF<double>());
+            else if (mdf == "beckmann")
+                m_mdf.reset(new BeckmannMDF<double>());
+            else return false;
+
+            return true;
+        }
+
+        virtual size_t compute_input_data_size(
+            const Assembly&         assembly) const APPLESEED_OVERRIDE
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Vertically align arguments.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+        APPLESEED_FORCE_INLINE virtual void sample(
+            SamplingContext&    sampling_context,
+            const void*         data,
+            const bool          adjoint,
+            const bool          cosine_mult,
+            BSDFSample&         sample) const APPLESEED_OVERRIDE
+        {
+            const InputValues *values = reinterpret_cast<const InputValues*>(data);
+
+            BackfacingPolicy backfacing_policy(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Make const?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+            const double cos_ih = dot(h, incoming);
+            const double cos_oh = dot(h, outgoing);
+
+            const double sqrt_denom = ior_i * cos_ih + ior_o * cos_oh;
+            if (sqrt_denom == 0.0)
+                return 0.0;
+
+            return
+                abs((cos_ih * cos_oh) / (cos_in * cos_on)) *
+                (square(ior_o) / square(sqrt_denom));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: `* square(ior_o / sqrt_denom)`

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+        {
+            // [1] eq. 17
+            const double cos_ih = dot(h, incoming);
+            const double cos_oh = dot(h, outgoing);
+
+            const double sqrt_denom = ior_i * cos_ih + ior_o * cos_oh;
+            if (sqrt_denom == 0.0)
+                return 0.0;
+
+            return (abs(cos_oh) * square(ior_o)) / square(sqrt_denom);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: `return abs(cos_oh) * square(ior_o / sqrt_denom)`

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.h [hunk 7]
```diff
@@ -46,31 +51,59 @@ namespace renderer
...
+    double      m_surface_transmittance_multiplier;
+    Spectrum    m_reflection_tint;
+    Spectrum    m_refraction_tint;
+    double      m_roughness;
+    double      m_anisotropic;
+    double      m_ior;
+    Spectrum    m_volume_transmittance;
+    double      m_volume_transmittance_distance;
+
+    // precomputed values.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.h]: *Precomputed*

File: src/appleseed/renderer/modeling/bsdf/glossybrdf.h [hunk 8]
```diff
@@ -61,6 +61,9 @@ APPLESEED_DECLARE_INPUT_VALUES(GlossyBRDFInputValues)
     double      m_roughness;
     double      m_anisotropic;
     double      m_ior;
+
+    // precomputed values.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossybrdf.h]: *Precomputed*

File: src/appleseed/renderer/modeling/material/oslmaterial.cpp [hunk 9]
```diff
@@ -30,7 +30,7 @@
 #include "oslmaterial.h"
 
 // appleseed.renderer headers.
-#include "renderer/modeling/bsdf/osl/oslbsdf.h"
+#include "renderer/modeling/bsdf/oslbsdf.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/material/oslmaterial.cpp]: Order alphabetically.

**Comment by dictoon** [comment]
Awesome! Good to merge.

The only thing I haven't quite understood is the `m_num_ior` thing. Also the nested media reference image should probably not be committed until absorption in the specular BTDF is fixed.

File: src/appleseed/renderer/kernel/shading/closures.cpp (hunk 2 above)
Comment (est77) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: It should be increased here, but I forgot do it. I'll fix it later today.

**Comment by dictoon** [comment]
Thanks!

**Pull request merged**